cli/packages.go

```diff
@@ -0,0 +1,135 @@
+	PackageTypeArchive     = PackageType("archive")
+	PackageTypeMLModel     = PackageType("ml_model")
+	PackageTypeModule      = PackageType("module")
+	PackageTypeSLAMMap     = PackageType("slam_map")
+)
+
+var packageTypes = []string{
+	string(PackageTypeUnspecified), string(PackageTypeArchive), string(PackageTypeMLModel),
+	string(PackageTypeModule), string(PackageTypeSLAMMap),
+}
+
+// PackageExportAction is the corresponding action for 'package export'.
+func PackageExportAction(c *cli.Context) error {
+	client, err := newViamClient(c)
+	if err != nil {
+		return err
+	}
+
+	return client.packageExportAction(
+		c.String(generalFlagOrgID),
+		c.String(packageFlagName),
+		c.String(packageFlagVersion),
+		c.String(packageFlagType),
+		c.Path(packageFlagDestination),
+	)
+}
+
+func (c *viamClient) packageExportAction(orgID, name, version, packageType, destination string) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+	// Package ID is the <organization-ID>/<package-name>
+	packageID := path.Join(orgID, name)
+
+	// Convert PackageType to proto
+	var packageTypeProto packagespb.PackageType
+	switch PackageType(packageType) {
+	case PackageTypeUnspecified:
+		packageTypeProto = packagespb.PackageType_PACKAGE_TYPE_UNSPECIFIED
+	case PackageTypeArchive:
+		packageTypeProto = packagespb.PackageType_PACKAGE_TYPE_ARCHIVE
+	case PackageTypeMLModel:
+		packageTypeProto = packagespb.PackageType_PACKAGE_TYPE_ML_MODEL
+	case PackageTypeModule:
+		packageTypeProto = packagespb.PackageType_PACKAGE_TYPE_MODULE
+	case PackageTypeSLAMMap:
+		packageTypeProto = packagespb.PackageType_PACKAGE_TYPE_SLAM_MAP
+	default:
+		return errors.New("invalid package type " + packageType)
+	}
+
+	resp, err := c.packageClient.GetPackage(c.c.Context,
+		&packagespb.GetPackageRequest{
+			Id:         packageID,
+			Version:    version,
+			Type:       &packageTypeProto,
+			IncludeUrl: &boolTrue,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return downloadPackageFromURL(c.c.Context, c.authFlow.httpClient, destination, name, version, resp.GetPackage().GetUrl())
+}
+
+func downloadPackageFromURL(ctx context.Context, httpClient *http.Client,
+	destination, name, version, packageURL string,
+) error {
+	// All packages are stored as .tar.gz
+	packagePath := filepath.Join(destination, version, name+".tar.gz")
+	if err := os.MkdirAll(filepath.Dir(packagePath), 0o700); err != nil {
+		return err
+	}
+	//nolint:gosec
+	packageFile, err := os.Create(packagePath)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, packageURL, nil)
+	if err != nil {
+		return errors.Wrapf(err, serverErrorMessage)
+	}
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return errors.Wrapf(err, serverErrorMessage)
+	}
+	if res.StatusCode != http.StatusOK {
+		return errors.New(serverErrorMessage)
+	}
+	defer func() {
+		utils.UncheckedError(res.Body.Close())
+	}()
+
+	bin, err := io.ReadAll(res.Body)
```

Comment:
IIUC this will read the entire package into RAM. I think this is a bit unexpected for the user, who might have enough space on disk for the file but not in RAM. I think the correct way to do this is to read in the file + write to disk in chunks. I did something like this in the file merge manager:

```golang
for {
	_, err := io.CopyN(localFileWriter, bigFileReader, inMemoryBufferSize)
	if err != nil {
		if errors.Is(err, io.EOF) {
			break
		}
		return err
	}
}
```
ChatGPT is suggesting using bufio.NewReader, which may be better.

Reply:
Thanks for the suggestion, updated!